File: standard-telegram-bot/src/main/java/com/immmus/standard/telegram/bot/WebHookTelegramBotService.java
package com.immmus.standard.telegram.bot;

import com.immmus.standard.telegram.bot.settings.WebHookBotSettings;
import lombok.extern.slf4j.Slf4j;
import org.telegram.telegrambots.bots.DefaultBotOptions;
import org.telegram.telegrambots.bots.TelegramWebhookBot;
import org.telegram.telegrambots.meta.api.methods.BotApiMethod;
import org.telegram.telegrambots.meta.api.objects.Update;

@Slf4j
final public class WebHookTelegramBotService extends TelegramBotService<TelegramWebhookBot> {
    private final TelegramWebhookBot client;

    public WebHookTelegramBotService(WebHookBotSettings settings, DefaultBotOptions defaultBotOptions) {
        this.client = new TelegramBot(settings, defaultBotOptions);
    }

    public WebHookTelegramBotService(WebHookBotSettings settings) {
        this(settings, settings.defaultBotOptions());
    }

    @Override
    public TelegramWebhookBot client() {
        return this.client;
    }

    @Override
    public void close() {
        log.info("{} is terminated", getClass().getSimpleName());
    }

    private class TelegramBot extends TelegramWebhookBot {
        private final String botUsername;
        private final String token;
        private final String webHookPath;

        public TelegramBot(WebHookBotSettings settings, DefaultBotOptions defaultBotOptions) {
            super(defaultBotOptions);
            this.botUsername = settings.botUsername();
            this.token = settings.token();
            this.webHookPath = settings.webHookPath();
        }

        @Override
        public BotApiMethod<?> onWebhookUpdateReceived(Update update) {
            BotApiMethod<?> result = update(update).orElse(null);
            log.debug("Update: {}. Message: {}. Successfully sent", update, result);
            return result;
        }

        @Override
        public String getBotUsername() {
            return this.botUsername;
        }

        @Override
        public String getBotToken() {
            return this.token;
        }

        @Override
        public String getBotPath() {
            return this.webHookPath;
        }
    }
}

File: standard-bot/src/main/kotlin/com/immmus/common/telegram/bot/WebHookTelegramBotService.kt
package com.immmus.common.telegram.bot

import com.immmus.common.telegram.bot.settings.WebHookBotSettings
import org.telegram.telegrambots.bots.DefaultBotOptions
import org.telegram.telegrambots.bots.TelegramWebhookBot
import org.telegram.telegrambots.meta.api.interfaces.BotApiObject
import org.telegram.telegrambots.meta.api.methods.BotApiMethod
import org.telegram.telegrambots.meta.api.objects.Message
import org.telegram.telegrambots.meta.api.objects.Update

abstract class WebHookTelegramBotService(
    settings: WebHookBotSettings,
    defaultBotOptions: DefaultBotOptions? = settings.defaultBotOptions()
) : TelegramBotService<TelegramWebhookBot>() {
    private val client: TelegramWebhookBot

    init {
        client = TelegramBot(settings, defaultBotOptions)
    }

    override fun client(): TelegramWebhookBot = client

    override fun close() {
        log.info("{} is terminated", javaClass.simpleName)
    }

    private inner class TelegramBot(
        private val settings: WebHookBotSettings,
        defaultBotOptions: DefaultBotOptions?
    ) :
        TelegramWebhookBot(defaultBotOptions) {

        override fun onWebhookUpdateReceived(update: Update): BotApiMethod<out BotApiObject>? {
            val result = update(update)
            log.debug("Update: {}. Message: {}. Successfully sent", update, result)
            return result
        }

        override fun getBotUsername(): String = settings.botUsername()

        override fun getBotToken(): String = settings.token()

        override fun getBotPath(): String = settings.webHookPath
    }
}